test/pipelinefinally_test.go

```diff
@@ -115,6 +142,27 @@ func TestPipelineLevelFinally_OneDAGTaskFailed_Failure(t *testing.T) {
 				t.Errorf("Error waiting for TaskRun to succeed: %v", err)
 			}
 			finalTaskStartTime = taskrunItem.Status.StartTime
+		case n == "finaltask2":
+			if err := WaitForTaskRunState(ctx, c, taskrunItem.Name, TaskRunSucceed(taskrunItem.Name), "TaskRunSuccess"); err != nil {
```

Comment:
I'm curious about the WaitForTaskRunState call here and in some of the other case blocks.  The test waits for the PipelineRun to fail (WaitForPipelineRunState) prior to this.  Can any TaskRuns still be running?  I know there have been problems in the past putting the PipelineRun into a done state before the TaskRuns were done but I thought that was resolved...

Reply:
Right, its waiting for the pipelineRun to fail using this function `WaitForPipelineRunState` right after creating a run. It's more of a polling for that state i.e. wait until pipeline is done. The state right away might not be set as expected. How can we make sure that the run finishes without waiting for it?

Reply:
I was saying that it's odd that after waiting for the pipelinerun to finish, the test also then waits for taskruns to finish.  The pipelinerun shouldn't be finished if there are taskruns still running.  I know there are some corner cases where this isn't true -- cancel comes to mind -- but I'm uncertain whether there is something going on here that makes it happen.

Reply:
Aah, I see what you mean. I am not sure either, pipelinerun completion mean all taskruns completed within the test `context`, ideally it should.

But, I see it as more than just waiting for the taskrun to finish, `WaitForTaskRunState` also verifies the state and returns error if state does not match. Do you suggest any alternative? I can follow up in a separate PR, may be write a separate routine to just validate the state?